app/src/main/kotlin/com/wire/android/ui/home/settings/backup/BackupAndRestoreViewModel.kt

```diff
@@ -1,17 +1,272 @@
+import com.wire.kalium.logic.util.fileExtension
 import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.delay
 import kotlinx.coroutines.launch
+import okio.Path
 import javax.inject.Inject
 
 @HiltViewModel
 class BackupAndRestoreViewModel
-@Inject constructor(private val navigationManager: NavigationManager) : ViewModel() {
+@Inject constructor(
+    private val navigationManager: NavigationManager,
+    private val importBackup: RestoreBackupUseCase,
+    private val createBackupUseCase: CreateBackupUseCase,
+    private val extractCompressedFileUseCase: ExtractCompressedBackupUseCase,
+    private val fileManager: FileManager,
+    private val kaliumFileSystem: KaliumFileSystem,
+    private val context: Context
+) : ViewModel() {
+
+    var state by mutableStateOf(BackupAndRestoreState.INITIAL_STATE)
+    private var latestCreatedBackup: BackupAndRestoreState.CreatedBackup? = null
+    private lateinit var latestExtractedBackupRootPath: Path
+
+    @Suppress("MagicNumber")
+    fun createBackup(password: String) {
+        viewModelScope.launch {
+            // TODO: Find a way to update the create progress more faithfully. For now we will just show this small delays to mimic the
+            //  progress also for small backups
+            state = state.copy(backupCreationProgress = BackupCreationProgress.InProgress(0.25f))
+            delay(300)
+            state = state.copy(backupCreationProgress = BackupCreationProgress.InProgress(0.50f))
+            delay(300)
+
+            when (val result = createBackupUseCase.invoke(password)) {
+                is CreateBackupResult.Success -> {
+                    state = state.copy(backupCreationProgress = BackupCreationProgress.Finished)
+                    latestCreatedBackup = BackupAndRestoreState.CreatedBackup(
+                        result.backupFilePath,
+                        result.backupFileName,
+                        result.backupFileSize,
+                        password.isNotEmpty()
+                    )
+                }
+                is CreateBackupResult.Failure -> {
+                    state = state.copy(backupCreationProgress = BackupCreationProgress.Failed)
+                    appLogger.e("Failed to create backup: ${result.coreFailure}")
+                }
+            }
+        }
+    }
+
+    fun saveBackup() = viewModelScope.launch {
+        latestCreatedBackup?.let { backupData ->
+            fileManager.shareWithExternalApp(backupData.path, backupData.assetName.fileExtension()) {}
+        }
+        state = BackupAndRestoreState.INITIAL_STATE
+    }
+
+    fun chooseBackupFileToRestore(uri: Uri) = viewModelScope.launch {
+        val importedBackupPath = kaliumFileSystem.tempFilePath(TEMP_IMPORTED_BACKUP_FILE_NAME)
+        uri.copyToTempPath(context, importedBackupPath)
+
+        extractBackupFiles(importedBackupPath)
+
+        // Delete the imported backup file
```

Comment:
```suggestion
```
Nothing blocking here, but I think we can remove this comment since it is self explanatory what's doing :D